04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+        },
+        "id": "4OJQgSzO4wCm",
+        "outputId": "605723f6-b536-400e-e9de-ae55f332b76e"
+      },
+      "execution_count": 31,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "<generator object <genexpr> at 0x7f2ca22f6950>\n",
+            "0\n",
+            "1\n",
+            "4\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "В горния пример, нашия генератор ще ни върне първите 5 точни квадрата, с помощта на generator expression. Как би изглеждал кода обаче, ако искаме да можем да приемаме `n`, с което да заместим 5 ?"
+      ],
+      "metadata": {
+        "id": "EhtFBeLX5i7B"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "perfect_squares_n_generator_exp = lambda n: (generate_nth_perfect_square(i) for i in range(1, n + 1))\n",
+        "\n",
+        "for num in perfect_squares_n_generator_exp(3):\n",
+        "    print(num)\n",
+        "\n",
+        "print('--')\n",
+        "for num in perfect_squares_n_generator_exp(10):\n",
+        "    print(num)"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "z2APkX8t5zj4",
+        "outputId": "452bc318-c244-48c5-bfe5-2e8e3972ea1e"
+      },
+      "execution_count": 32,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "0\n",
+            "1\n",
+            "4\n",
+            "--\n",
+            "0\n",
+            "1\n",
+            "4\n",
+            "9\n",
+            "16\n",
+            "25\n",
+            "36\n",
+            "49\n",
+            "64\n",
+            "81\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Едно последно нещо - един от най-използваните генератори е `range`. `range` приема три аргумента - начало, край и стъпка. В резултат, получаваме генератор, който генерира числата от това множество. "
+      ],
+      "metadata": {
+        "id": "G-OM7Qoi8AoH"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "for i in range(2, 7, 2):\n",
+        "    print(i)"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "aT8UCA-L8jty",
+        "outputId": "61db65bb-5dd9-4a8f-a63e-abb263f11aa0"
+      },
+      "execution_count": 33,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "2\n",
+            "4\n",
+            "6\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Няма значение колко е голям размера на `range`-а за представянето му - това е защото `range` изчислява следващия елемент на момента. "
```

Comment:
If we'll be comparing python 2 and 3, this is a good place to mention that python 2 range is not lazy, and there is a lazy version `xrange`

Reply:
Will mention it in the Python 2 vs 3 lecture